src/Silverback.Integration/Messaging/Configuration/EndpointsConfiguratorsInvoker.cs

```diff
@@ -37,6 +37,9 @@ public void Invoke()
 
                     scope.ServiceProvider.GetServices<IEndpointsConfigurator>()
                         .ForEach(configurator => configurator.Configure(endpointsConfigurationBuilder));
+
+                    IBrokerEvents brokerEvents = scope.ServiceProvider.GetRequiredService<IBrokerEvents>();
+                    brokerEvents.OnEndpointsConfigured?.Invoke(scope.ServiceProvider);
```

Comment:
We discussed to pass the endpoints as argument to the callback. But I was not sure how to get them and whether it actually makes sense.
I pass the `ServiceProvider` for now, but we should probably change this to some other argument because it gives an incentive to use ServiceLocator pattern (which is not good).

Reply:
You will call the `WithConnectionToMessageBroker` and therefore the `WithEvents` method from the `ConfigureServices` method and you don't have a convenient way to retrieve an instance of the `IServiceProvider` or resolve any scoped service otherwise.  But yeah, I don't like to pass the `IServiceProvider` as argument to the callback.

In this case I believe that we should use a different approach that what we do in the endpoints and have an actual type (much like the `IEndpointsConfigurator`) that declares the callback methods and be registered as service and resolved with all its dependencies.

It could be something like this:

```csharp
public interface IBrokerEventsHandler
{
    Task OnEndpointsConfigured(IReadOnlyCollection<IEndpoint> endpoints);
}
```

Then there could be a base class, for convenience. So that one isn't forced to declare all handlers, if we ever add other events.

```csharp
public abstract class BrokerEventsHandler : IBrokerEventsHandler
{
    public virtual Task OnEndpointsConfigured(IReadOnlyCollection<IEndpoint> endpoints) => Task.CompletedTask;
}
```

And at that point have the configuration method like `AddEventHandler<T>` instead of `WithEvents`.

It's a bit different but I think it's much better for the end user, since you almost for sure need some references to be resolved and passing the `IServiceProvider` around isn't that elegant.

Reply:
About the other question, how to retrieve the list of endpoints.

I believe that you could simply enumerate the consumers from the `IBrokerCollection` and read the producer endpoints from the `IOutboundRoutingConfiguration`, like you did to prevent the double registration.

Consumer endpoints:
```
var endpoints  = brokerCollection
    .SelectMany(broker => broker.Consumers)
    .Select(consumer => consumer.Endpoint)
    .ToList();
```

Producer endpoints:
```csharp
var endpoints = endpointsConfigurationBuilder
    .GetOutboundRoutingConfiguration()
    .Routes.Where(r => r.MessageType == messageType)
    .Select(route => route.GetOutboundRouter(endpointsConfigurationBuilder.ServiceProvider))
    .OfType<StaticOutboundRouter>()
    .SelectMany(router => router.Endpoints)
    .ToList();
```

(You could use the `IBrokerCollection` to enumerate the producers as well, but that would only work if those have been preloaded...that's actually the default, but still can be disabled.)